### DISCUSSION TITLE
Join across data sources - how?

### DISCUSSION THREAD

**Discussion by stevenhorner** [neilotoole/sq#440] [Q&A]
I have read the documentation joins dozens of times but I can't get the join across data sources to work between 2 SQL Server DBs. I have both sources added and everything worked on each source individually as expected but if I try a join between 2 data sources it only ever returns all records and all fields from the left side.

This is an example:

`sq '@dev.hdSettings:d | join(@preprod.hdSettings:p, .d.InstanceID == .p.InstanceID) | .d.Title, .p.AppURL'`

I've tried many variations on this but no matter what I do I can not get a join between the 2 SQL Servers to work. In the example above there is only one record and both of them have an InstanceID of `0`.

Here are some other examples, all just return all records and fields from the left side, they don't even only show the specified fields:
`sq '@preprod.hdSettings:a, @dev.hdSettings:b | join(.InstanceID) | .a.Title'`

`sq '@preprod.hdSettings:a | join(@dev.hdSettings:b, .InstanceID) | .InstanceID, .a.Title, .b.Title'`

What am I doing wrong, the documentation gives examples of cross joining multiple data sources of different types of DB?

Also noticed that if I do a diff across the 2 data sources, the first specified @dev one below is only shown:
`sq diff @dev.hdSettings @preprod.hdSettings --data`

**Comment by neilotoole** [comment]
@stevenhorner: Thanks for the detailed report, and sorry it sat for so long. The join issue you hit is the exact bug fixed by [#445](https://github.com/neilotoole/sq/issues/445), shipped in [**v0.53.0**](https://github.com/neilotoole/sq/releases/tag/v0.53.0) (2026-05-25):

> Cross-source `join` no longer fails when the participating sources contain tables with the same name (e.g. `@src1.actor | join(@src2.actor, .actor_id)`).

Your case is the same shape — `hdSettings` exists on both `@dev` and `@preprod`, and the planner was previously collapsing the right-hand source's table reference onto the left. After upgrading, your original query should work:

```shell
sq '@dev.hdSettings:d | join(@preprod.hdSettings:p, .d.InstanceID == .p.InstanceID) | .d.Title, .p.AppURL'
```

See the [join docs](https://sq.io/docs/query#join) for the cross-source same-name rename behavior.

Separately, the `sq diff @dev.hdSettings @preprod.hdSettings --data` observation looks like a different problem (diff is a separate code path from the SLQ planner). If it's still misbehaving on v0.53.0, please open a fresh issue with a reproducer and I'll take a look.
